src/init.c

```diff
@@ -686,207 +696,92 @@ static int hyper_cmd_write_file(char *json, int length)
-		uint32_t type;
-
-		if (hyper_get_type(pipe[0], &type) < 0 || type != READY) {
-			fprintf(stderr, "get incorrect message type %d, expect READY\n", type);
-			goto out;
-		}
-
-		ret = 0;
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, sockpair) < 0) {
+		perror("fail to set socket pair");
 		goto out;
 	}
 
-	/* TODO: wait for container finishing setup root */
-	if (setns(mntns, CLONE_NEWNS) < 0) {
-		perror("fail to enter container ns");
-		goto exit;
-	}
-
-	fprintf(stdout, "write file %s, data len %d\n", cmd.file, datalen);
-
-	fd = open(cmd.file, O_CREAT| O_TRUNC| O_WRONLY, 0644);
-	if (fd < 0) {
-		perror("fail to open target file");
-		goto exit;
-	}
-
-	while(len < datalen) {
-		size = write(fd, data + len, datalen - len);
-
-		if (size < 0) {
-			if (errno == EINTR)
-				continue;
-
-			perror("fail to write data to file");
+	pid = fork();
+	if (pid < 0) {
+		perror("fail to fork child process");
+		goto out;		
+	} else if (pid == 0) {
+		/* TODO: wait for container finishing setup root */
+		if (setns(mntns, CLONE_NEWNS) < 0) {
+			perror("fail to enter container ns");
 			goto exit;
 		}
 
-		len += size;
-	}
-	ret = 0;
-exit:
-	hyper_send_type(pipe[1], ret ? ERROR : READY);
-	_exit(0);
-out:
-	close(pipe[0]);
-	close(pipe[1]);
-	free(cmd.id);
-	free(cmd.file);
-
-	return 0;
-}
-
-struct hyper_file_arg {
-	int		mntns;
-	int		pipe[2];
-	char		*file;
-	char		root[128];
-	uint32_t	*datalen;
-	uint8_t		**data;
-};
-
-static int hyper_do_cmd_read_file(void *data)
-{
-	struct stat st;
-	int len = 0, size, fd, ret = -1;
-	struct hyper_file_arg *arg = data;
-
-	/* TODO: wait for container finishing setup root */
-	if (setns(arg->mntns, CLONE_NEWNS) < 0) {
-		perror("fail to enter container ns");
-		goto err;
+		if (rw == WRITEFILE) {
+			fd = open(cmd.file, O_CREAT| O_TRUNC| O_WRONLY, 0644);
+		} else {
+			fd = open(cmd.file, O_RDONLY);
+		}
+		if (fd < 0) {
+			perror("fail to open target file");
+			goto exit;
+		}
+		if (hyper_send_fd(sockpair[0], fd) < 0) {
+			perror("fail to send fd");
+			goto exit;
+		}
+		close(fd);
 	}
 
-	fprintf(stdout, "read file %s\n", arg->file);
-
-	if (stat(arg->file, &st) < 0) {
-		perror("fail to state file");
-		goto err;
+	if (hyper_recv_fd(sockpair[1], &fd) < 0) {
+		perror("fail to recv fd");
+		goto out;
 	}
 
-	*arg->datalen = st.st_size;
-	*arg->data = malloc(*arg->datalen);
-	if (*arg->data == NULL) {
-		fprintf(stderr, "allocate memory for reading file failed\n");
-		goto err;
+	if (rw == READFILE) {
+		if(fstat(fd, &st) < 0) {
+			perror("fail to state file");
+			goto out;
+		}
+		*rdatalen = datalen = st.st_size;
+		*rdata = data = malloc(datalen);
+		if (*rdata == NULL) {
+			fprintf(stderr, "allocate memory for reading file failed\n");
+			goto out;
+		}
+		fprintf(stdout, "file length %d\n", *rdatalen);
 	}
 
-	fd = open(arg->file, O_RDONLY);
-	if (fd < 0) {
-		perror("fail to open target file");
-		goto err;
+	if (rw == WRITEFILE) {
+		fileoprt = write;
+	} else {
+		fileoprt = read;
 	}
```

Comment:
don't need fileoprt. just use the `if (...) {read(...) } else { write(...); } ` in the while loop